### DISCUSSION TITLE
Boot Mode 2 (User Secure Boot)

### DISCUSSION THREAD

**Discussion by utku1998** [polarfire-soc/polarfire-soc-documentation#270] [General]
Hello,

I am currently working with the "PolarFire SoC FPGA Icicle Kit". I have configured the device with the newest "Reference Design".

I want to implement the full boot flow of the "Boot Mode = 2 (User Secure Boot)". According to the documentation [here](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/fundamentals/boot-modes/boot-mode-2/boot-mode-2-fundamentals.md), in this boot mode we flash our own code to the "sNVM". In the run time, the system controller authenticates the content of the "sNVM", copies it to "On-chip RAM" and executes from there.

In the documentation, it is also specified that after running the code in "sNVM", which typically authenticates the content of the "eNVM" before executing it.

So, I have 2 questions:

1) How can I program the "sNVM" and "eNVM" seperately? For instance, we can seperately program the "eNVM" with the option "Boot Mode = 1" and the "sNVM" with the option "Boot Mode = 2 (User Secure Boot)" using "PolarFire SoC/Libero." But each of them sets the boot mode to either 1 or 2. However, what I want to do is program both of them (eNVM and sNVM) using the Boot Mode = 2 (User Secure Boot).

2) Which configurations should I make that after the device executes the code in "sNVM", it moves on with the "eNVM"?

Best regards!

**Comment by TommyMurphyTM1234** [comment]
You should be able to create two programs - one for eNVM and one for sNVM. The former is the "main" program that you want to run. The latter is the boot mode 2 user secure bootloader that runs from startup after the normal boot mode 2 secure checks, does whatever init it needs to do, then transfers control to the "main" program in eNVM. When using SoftConsole you should be able to use the boot mode 1 support to get the "main" program into eNVM. (It will also configure the target for boot mode 1, but this is just temporary and will be overridden in the next step). Once that is done and the "main" program stored in eNVM, you can use the boot mode 2 support to get the bootloader into sNVM and the target configured for boot mode 2.

If using Libero/FlashPro etc. you configure the boot mode (2) there and the sNVM and eNVM program payloads in the relevant project and program them from there (most likely along with the hardware design - MSS configuration, FPGA bitfile etc.).

My recollection is also that the HSS (https://github.com/polarfire-soc/hart-software-services) may have its own way of booting from sNVM in user secure boot mode 2 and then transferring control to code stored in eNVM. So it may be worth having a look at the HSS in case its approach is more appropriate for your needs.

**Reply by utku1998** [reply]
@TommyMurphyTM1234 Thank you very much for your reply :). 

Is that what you mean?: So I will program with the "Libero" the "sNVM" (chosing the option "Add Boot Mode 2 Authenticated Client") and "eNVM" (chosing which option below?). In this case, will the "Boot Mode = 1" be overridden by the "Boot Mode = 2"? Beacuse I choose the "Boot Mode = 2" in "sNVM" configuration, while I choose the "Boot Mode = 1" option in eNVM configuration (please see the pictures below).

![image](https://user-images.githubusercontent.com/84975284/208931971-a3b5d608-e736-4624-8999-6e2e0ae78bf5.png)

![image](https://user-images.githubusercontent.com/84975284/208931833-2292eafd-f046-4826-acf3-fa0f88e3a7a5.png)

PS. I have looked to the HSS, but could not find/understand how it goes from eNVM to SDCard. Could you provide me a more specific section or code snippet (in HSS or other), if you have experience about it?

Best regards!

**Reply by TommyMurphyTM1234** [reply]
> @TommyMurphyTM1234 Thank you very much for your reply :).

You're welcome.

> Is that what you mean?: So I will program with the "Libero" the "sNVM" (chosing the option "Add Boot Mode 2 Authenticated Client")

If doing it from Libero, then, yes.

> and "eNVM" (chosing which option below?). In this case, will the "Boot Mode = 1" be overridden by the "Boot Mode = 2"? 

It's a while since I've used Libero but isn't there an option to configure an eNVM storage only (no boot mode 1 configuration) client for storing the "main" program which the sNVM boot mode 2 bootloader will transfer control to?

I don't think that Libero is going to allow you to configure a boot mode 1 eNVM client and a boot mode 2 sNVM client at the same time. I would expect some configuration time or compile time DRC to catch this as an error.

As I mentioned above, programming the boot mode 1 eNVM program followed by the boot mode 2 sNVM client from SoftConsole should leave the target in boot mode 2 such that the boot mode 2 bootloader starts and, assuming it's coded/constructed appropriately, then transfers control to the "main" program in eNVM.

> PS. I have looked to the HSS, but could not find/understand how it goes from eNVM to SDCard. Could you provide me a more specific section or code snippet (in HSS or other), if you have experience about it?

Again, it's a while since I played with the HSS but my understanding was that it generally or always booted from eNVM or sNVM and not from sdcard? Are you sure that you're not confusing the HSS with Linux (Yocto or Buildroot generated)?

**Reply by utku1998** [reply]
@TommyMurphyTM1234

> It's a while since I've used Libero but isn't there an option to configure an eNVM storage only (no boot mode 1 configuration) client for storing the "main" program which the sNVM boot mode 2 bootloader will transfer control to? 

No, unfortunately in the "eNVM" section, there are only these 2 options: 

![image](https://user-images.githubusercontent.com/84975284/208999993-7f717262-a640-4be4-98bd-ee3551bb9e27.png)

And in the boot mode 2, if we somehow successfully program the "eNVM" and "sNVM" either with "Libero" or "SoftConsole IDE" (which does not seem possible right now!): The device will firstly execute the code from "sNVM". And will it then automatically move on executing the code in "eNVM"? Or do I need to make a configuration or explicitly write a code and the end of the code in "sNVM", which then moves on executing the code in "eNVM"? How does this transition process (from sNVM to eNVM) works?

> As I mentioned above, programming the boot mode 1 eNVM program followed by the boot mode 2 sNVM client from SoftConsole should leave the target in boot mode 2 such that the boot mode 2 bootloader starts and, assuming it's coded/constructed appropriately, then transfers control to the "main" program in eNVM.

I must try it with "Libero", since there is a bug in "SoftConsole IDE" for the "Boot Mode = 2 (User Secure Boot)". I have also talked about it with the "Microchip" engineers and they will hopefully solve this issue in the next release of "SoftConsole". 

> Again, it's a while since I played with the HSS but my understanding was that it generally or always booted from eNVM or sNVM and not from sdcard? Are you sure that you're not confusing the HSS with Linux (Yocto or Buildroot generated)?

No, what I mean is; the HSS starts executing (boot up) from the eNVM and after it is finished executing, moves on with the payload (code) from the SD card.

**Reply by TommyMurphyTM1234** [reply]
> No, unfortunately in the "eNVM" section, there are only these 2 options:

Ok, I thought that there would be an option to store data (actually program code) in an eNVM region without having to set the boot mode (to 1). Other Microchip SoC FPGAs have that option so I'm surprised that PolarFire SoC doesn't.

> And in the boot mode 2, if we somehow successfully program the "eNVM" and "sNVM" either with "Libero" or "SoftConsole IDE" (which does not seem possible right now!): The device will firstly execute the code from "sNVM". And will it then automatically move on executing the code in "eNVM"?

Not automatically.

> Or do I need to make a configuration or explicitly write a code and the end of the code in "sNVM", which then moves on executing the code in "eNVM"? How does this transition process (from sNVM to eNVM) works?

I would expect a simple jump/call to the relevant code in eNVM to do it. I can't see why anything else would be required?

> I must try it with "Libero", since there is a bug in "SoftConsole IDE" for the "Boot Mode = 2 (User Secure Boot)".

What bug?

> No, what I mean is; the HSS starts executing (boot up) from the eNVM and after it is finished executing, moves on with the payload (code) from the SD card.

Isn't that just one option and not necessarily a requirement when using the HSS?

**Reply by utku1998** [reply]
> Ok, I thought that there would be an option to store data (actually program code) in an eNVM region without having to set the boot mode (to 1). Other Microchip SoC FPGAs have that option so I'm surprised that PolarFire SoC doesn't.

Yeah, this is really interesting. And as I said, I can not do it with the "SoftConsole IDE" (since it is buggy for the Boot Mode = 2, although the Boot Mode = 0/1 works!). 

> I would expect a simple jump/call to the relevant code in eNVM to do it. I can't see why anything else would be required?

Do you mean directly changing the "PC" (Program Counter) to the memory address of the eNVM using "inline assembly" etc. (or to whereever the code starts executing in "eNVM"). I have searched for this "jump/call" instruction in the "HSS" (which may be helpful for my case to jump from the "sNVM" to "eNVM"), but unfortunately I could not find that "jump/call" instruction in the HSS. I would be happy to hear, if you have an idea for where it could be or how can I jump from the "sNVM" to "eNVM" in my case? :)

> What bug?

In my previous post (which you have also answered), I have explained detailed the error that I got, while I am trying the "Boot Mode = 2". And the engineers from the Microchip tried also the option "Boot Mode = 2" with "SoftConsole IDE" and they got the same problem. Then, they told me that there is a bug with the "SoftConsole IDE", which they will try to fix internally for the next release of the "SoftConsole IDE".

Here is my previous post and the problem that I have mentioned: 

"""
I am currently working with the "PolarFire SoC FPGA Icicle Kit". I have configured/flashed the device with the newest "Reference Design". Then, I have created an example test code (Simple LED Blinking example) in order test it for different boot modes (Boot Mode = 0, Boot Mode = 1, Boot Mode = 2, Boot Mode = 3). I have managed to boot the device with the Boot Modes = 0/1 and run the blinking code using the SoftConsole IDE. However I could not manage to boot the device in Boot Mode = 2 (User Secure Boot). While I am trying to flash my blinking code using SoftConsole IDEs Boot Mode = 2 option (with LIMRelease, which should be suitable for the User Secure Boot), I am getting the Error: "Bitstream or data is corrupted or noisy. ERROR_CODE: 801F". [Here](https://github.com/polarfire-soc/polarfire-soc-documentation/files/10219169/debugLog.txt), you can also see the "debugLog.txt".
"""

> Isn't that just one option and not necessarily a requirement when using the HSS?

Yes, I think it is just one option, which is explained in the documentation and videos of @hughbreslin.

**Reply by TommyMurphyTM1234** [reply]
> Yeah, this is really interesting. And as I said, I can not do it with the "SoftConsole IDE" (since it is buggy for the Boot Mode = 2, although the Boot Mode = 0/1 works!).

I didn't think that it's a SoftConsole bug. I linked to the relevant SoftConsole documentation in your other thread. The issue is almost certainly that the SMK (sNVM Master Key) needs to be (re)initialised via Libero before boot mode 2 programming from SoftConsole can succeed.

> Do you mean directly changing the "PC" (Program Counter) to the memory address of the eNVM using "inline assembly" etc. (or to whereever the code starts executing in "eNVM"). I have searched for this "jump/call" instruction in the "HSS" (which may be helpful for my case to jump from the "sNVM" to "eNVM"), but unfortunately I could not find that "jump/call" instruction in the HSS. I would be happy to hear, if you have an idea for where it could be or how can I jump from the "sNVM" to "eNVM" in my case? :)

Sort of - I mean something like a jump to the relevant eNVM address. Jump is a RISC-V (pseudo) instruction.

https://web.eecs.utk.edu/~smarz1/courses/ece356/notes/assembly/

> In my previous post (which you have also answered), I have explained detailed the error that I got, while I am trying the "Boot Mode = 2". And the engineers from the Microchip tried also the option "Boot Mode = 2" with "SoftConsole IDE" and they got the same problem. Then, they told me that there is a bug with the "SoftConsole IDE", which they will try to fix internally for the next release of the "SoftConsole IDE".

See above and in the other thread. I don't think it's a bug but rather that the SMK simply hasn't been initialised.

**Reply by utku1998** [reply]
> I didn't think that it's a SoftConsole bug. I linked to the relevant SoftConsole documentation in your other thread. The issue is almost certainly that the SMK (sNVM Master Key) needs to be (re)initialised via Libero before boot mode 2 programming from SoftConsole can succeed.

I do not know. But a Microchip engineer and his collegue have checked this "Boot Mode = 2" option for me in the "SoftConsole IDE". And they have told me that it does not work on them either and they get the same error. But for your solution, do you know how to (re)initiliaze the "SMK" key to "PolarFire SoC"? I do not see any option for that in "Libero" in the "sNVM" section (see below!). 

![image](https://user-images.githubusercontent.com/84975284/209111190-da8316e9-7ce0-4307-b95d-dbc902675737.png)


> Sort of - I mean something like a jump to the relevant eNVM address. Jump is a RISC-V (pseudo) instruction.

If it is the way to jump from one memory element to another (e.g. from "eNVM" to "sNVM" in my case, and from "eNVM" to "SD Card" for the case of "HSS"), where has this "jump/call" done in "HSS". I could not figure it out in the source code of the "HSS". Can you guide me, if you could see that place in the source code (https://github.com/polarfire-soc/hart-software-services). If I can find it, I can simply use the same code/instruction for my case.

**Reply by TommyMurphyTM1234** [reply]
> I do not know. But a Microchip engineer and his collegue have checked this "Boot Mode = 2" option for me in the "SoftConsole IDE". And they have told me that it does not work on them either and they get the same error. But for your solution, do you know how to (re)initiliaze the "SMK" key to "PolarFire SoC"? I do not see any option for that in "Libero" in the "sNVM" section (see below!).

You need to create a simple design in Libero with security enabled and when that's programmed it will (re)initialise the SMK. You'll then need to reprogram it with the security options disabled unless you're actively using them. This will leave the SMK as it was and boot mode 2 programming from SoftConsole should then work. I presume that the security options and SMK initialisation process are covered in the Libero documentation.

> If it is the way to jump from one memory element to another (e.g. from "eNVM" to "sNVM" in my case, and from "eNVM" to "SD Card" for the case of "HSS"), where has this "jump/call" done in "HSS". I could not figure it out in the source code of the "HSS". Can you guide me, if you could see that place in the source code (https://github.com/polarfire-soc/hart-software-services). If I can find it, I can simply use the same code/instruction for my case.

I'm not really sure where the confusion arises. It's simply a jump to the relevant address.

**Reply by utku1998** [reply]
> You need to create a simple design in Libero with security enabled and when that's programmed it will (re)initialise the SMK. You'll then need to reprogram it with the security options disabled unless you're actively using them. This will leave the SMK as it was and boot mode 2 programming from SoftConsole should then work. I presume that the security options and SMK initialisation process are covered in the Libero documentation.

You mean firsty, choosing the option "custom security options" (see picture below). This will automatically initiliaze the "SMK"? And then what you mean with disabling is: creating a new design and configuring the bord with the opton "Bitstream Encryption with the Defaul Key" right? Otherwise, we can not reprogram the device using "SoftConsole IDE", since it has not the keys that we have initialized, while choosing the option "custom security options". So, we need to disable it first.

![image](https://user-images.githubusercontent.com/84975284/209122489-3f33b5ca-4cfa-443e-84fd-42953bed581e.png)

> I'm not really sure where the confusion arises. It's simply a jump to the relevant address.

I have understood the idea of "jump" from the "sNVM" to "eNVM". But I could not find that "jump" instruction/C-code in the source of the HSS (https://github.com/polarfire-soc/hart-software-services). So, in the case of HSS, there should be also a "jump" from the "eNVM" to "SD Card" or "eMMC". But, the question is: where is it?

**Reply by TommyMurphyTM1234** [reply]
> You mean firsty, choosing the option "custom security options" (see picture below). This will automatically initiliaze the "SMK"? And then what you mean with disabling is: creating a new design and configuring the bord with the opton "Bitstream Encryption with the Defaul Key" right? Otherwise, we can not reprogram the device using "SoftConsole IDE", since it has not the keys that we have initialized, while choosing the option "custom security options". So, we need to disable it first.

Something like that, but you'll need to consult the Libero documentation for the exact details of how to enable security (so that the SMK gets (re)initialised) and how to then disable security leaving the SMK set and the target accessible from SoftConsole for boot mode 2 programming. You are correct that if security is left enabled then SoftConsole debugging/boot mode programming will be locked out.

> I have understood the idea of "jump" from the "sNVM" to "eNVM". But I could not find that "jump" instruction/C-code in the source of the HSS (https://github.com/polarfire-soc/hart-software-services). So, in the case of HSS, there should be also a "jump" from the "eNVM" to "SD Card" or "eMMC". But, the question is: where is it?

Sorry, I don't know about the HSS hand-off to non-bootloader code, but the fact remains that if you want to start executing code in eNVM you should be able to just jump to it or use a predefined symbol with the correct address to call into it.
